### ISSUE TITLE
Text become messed up after toggling Texture Upscaling ON/OFF

### ISSUE THREAD

**Issue by bennywan24** [hrydgard/ppsspp#8797]
In the latest version of PPSSPP (v 1.2.2 ; March 19, 2016), text would becomes messed up if the Texture Upscaling option was to be toggled when in mid-game (particularly Turn OFF Texture Upscaling). This issue occurs even with every other settings at default.

The game which I found this problem is at:  **Corpse Party: Book of Shadows (USA)** . 
A screenshot of this glitch can be seen here:

When using OpenGL:
[http://imgur.com/u9bvDkO](http://imgur.com/u9bvDkO)
[http://imgur.com/LEN5WNg](http://imgur.com/LEN5WNg)

When using Direct3D9:
[http://imgur.com/KeXSY6S](http://imgur.com/KeXSY6S)

_PPSSPP Version:_
64 bits

_Computer:_
Windows 10

I have tried even replacing the default font of PPSSPP with the official PSP ones, but it doesn't seem to make a difference...

**Comment by unknownbrackets** [comment]
Strange.  Can you try the latest git build?

http://buildbot.orphis.net/ppsspp/

We've made some changes to the way textures are handled when upscaling.

-[Unknown]

**Comment by bennywan24** [comment]
Just tested the latest git build, the issues is still unchanged :(

However, I also found out that the graphical glitches disappear after resizing the display, so it seems like it isn't that big of issue.

**Comment by unknownbrackets** [comment]
Does anything show up in the error log?  Does it happen consistently every time?

I'm not seeing this in other games, but I don't have that specific game.

Does it happen only when changing via the Windows menu, or does it also happen when changing from the pause menu?

-[Unknown]

**Comment by bennywan24** [comment]
- Does anything show up in the error log?
Can you tell me how I would access the error log? 

- Does it happen consistently every time?
Though the exact area of graphical glitches seem to change at times (i.e. covering different letters etc.); it is bounded to occur after a few pagea of text changes almost everytime; provided that Texture Upscaling was ON beforehand and was switched to OFF thru the _pause_ menu.

- Does it happen only when changing via the Windows menu, or does it also happen when changing from the pause menu?
Interestingly enough, this issue **ONLY** occur when using the PAUSE menu. Using the Window menu bars seems to be perfectly ok.

**Comment by unknownbrackets** [comment]
The log - I think the buildbot comes with a batch file that will create a log, but you can also access it from the Debug menu (Log Console.)

Interesting that it's through pause only.  I was thinking maybe it was a threading thing, but it can't be in that case.

Does it ever happen on the text displayed on screen after you unpause?  Or only when the game draws more text?

Hmm.  Perhaps, does it help to turn off vertex cache?  Meaning, disable vertex cache, play a bit, and then switch texture upscaling off.  I doubt that's it, though...

-[Unknown]

**Comment by bennywan24** [comment]
- I have taken a look through the log, and found some errors (highlighted in RED/YELLOW):

01:40:705 user_main    E[IO]: HLE\sceIo.cpp:1730 UNIMPL sceIoChangeAsyncPriority(4, 24)
01:40:707 user_main    E[IO]: HLE\sceIo.cpp:1730 UNIMPL sceIoChangeAsyncPriority(4, 24)
01:40:708 user_main    E[IO]: HLE\sceIo.cpp:1730 UNIMPL sceIoChangeAsyncPriority(5, 24)
01:40:740 CRI ADX File E[IO]: HLE\sceIo.cpp:1730 UNIMPL sceIoChangeAsyncPriority(6, 24)
01:40:818 CRI ADX File E[IO]: HLE\sceIo.cpp:1730 UNIMPL sceIoChangeAsyncPriority(6, 24)
01:42:838 CRI ADX File E[IO]: HLE\sceIo.cpp:1730 UNIMPL sceIoChangeAsyncPriority(6, 24)
01:47:767 SofdecThread W[ME]: HLE\sceMpeg.cpp:1754 UNIMPL sceMpegFlushAllStream(08957d70)
01:47:768 SofdecThread E[ME]: HLE\sceMpeg.cpp:1285 UNIMPL sceMpegAvcDecodeFlush(08957d70)
01:47:866 CRI ADX File E[IO]: HLE\sceIo.cpp:1730 UNIMPL sceIoChangeAsyncPriority(6, 24)
01:50:251 SofdecThread W[ME]: HLE\sceMpeg.cpp:1754 UNIMPL sceMpegFlushAllStream(08957d70)
01:50:251 SofdecThread E[ME]: HLE\sceMpeg.cpp:1285 UNIMPL sceMpegAvcDecodeFlush(08957d70)
06:18:291 CRI ADX Audi E[ME]: HLE\sceAtrac.cpp:562 avcodec_decode_audio4: Error decoding audio -1094995529 / bebbb1b7


- Does it ever happen on the text displayed on screen after you unpause? Or only when the game draws more text?

The glitch will only manifest after some new text is drawn.

- Hmm. Perhaps, does it help to turn off vertex cache?

Interestingly, disabling Vertex Cache makes the problem worse! Doing so render every text messed up, where as before only some are messed up.

**Comment by unknownbrackets** [comment]
Ahhh... do you have "Texture coord speedhack (speedup)" unchecked?  That's on my default and not really any more of a hack than Software Skinning (right, @hrydgard?)  Maybe we should move it.

Perhaps there's an issue when that's unchecked, that could be... we might be detecting a max V, and this game might be using tex scale/offset.

If the problem goes away with that option checked, I think the solution is just to make sure the min/max UV checks are only done when that option is enabled.

-[Unknown]

**Comment by bennywan24** [comment]
Hmm.... sadly, the "Texture coord speedhack (speedup)" option doesn't seem to affect the outcome much :(

Yeah,  "Texture coord speedhack (speedup)" is OFF by default.

**Comment by unknownbrackets** [comment]
Hmm, darn.  If disabling vertex cache makes it worse, it really makes me think that vertex cache was just hiding the problem - and that means it's somehow related to vertex decode.

We flush the texture cache when leaving the pause menu already... strange.

Starting to wonder if this might even be a driver bug, triggered by us clearing cache somehow.  It doesn't happen in any other games?

-[Unknown]

**Comment by bennywan24** [comment]
I found another game that is affected by this issue, though it looks like it is not as severe. 

**Jeanne d'Arc**:
http://imgur.com/BuSGaU6


Perhaps this issue only happens when the text of a game can be improved by the Texture Upscaling algorithm?

**Comment by unknownbrackets** [comment]
I have Jeanne d'Arc.  I must be doing something differently, since I didn't reproduce the issue after a few tries.  Here are my exact steps:

1. Set 3x internal render resolution, 3x texture hybrid scaling using deposterize.  Vertex cache off, frameskip off.  Everything else at defaults.

2. Start game and fast forward through intros.  Note: latest build has a bug and needs #8807.

3. At New Game, look at first dialog box.

4. Press Esc to open the menu.

5. Under Settings, disable texture upscaling by setting it to Off.  Change nothing else, and resume game play via clicking "Continue" with the mouse.

6. Watch text for a bit, then advance to another dialog box.  Re-enable upscaling, watch text, advance, and disable again.

Could you be using texture replacement at all?  I'm hoping #8806 is the cause, but I tested with and without that change and still could not reproduce the issue either way.  I did not have any saved PNG texture replacements, though.

If you are using the new replacement feature, you're almost definitely hitting the bug fixed in #8806.

-[Unknown]

**Comment by bennywan24** [comment]
So I followed your instructions and tested it. And the weird thing is, the problem didn't occur!
I believe that **Hybrid Scaling** might be _immune_ to this issue, as other Scaling methods still produce the glitch; though the glitch might not show all straight so away, so you might need to advance through a few textboxes.

Also, I don't believe I'm using texture replacement? I think I might have replaced the font used but that's about it.

**Comment by unknownbrackets** [comment]
If you don't know you're using texture replacement, you aren't.

I wonder if this is somehow some kind of race condition then?  Unfortunately, I'm still not reproducing it even with xBRZ.  And if you're using the latest git build, it sounds like #8806 didn't help after all.

Which kind of video card was this?  NVIDIA, AMD, Intel, or something else?  Maybe we're triggering a driver bug somehow.

-[Unknown]

**Comment by bennywan24** [comment]
Good news, I have just downloaded the latest build (v1.2.2-666) and turns out the issue have been fixed! I never realized how long behind the version I had before was lagging behind.
I don't know exactly what you guys did but good job on fixing it!

As for my video card, it's:
**Intel HD Graphics 4600** + NVIDIA Geforce GTX 860M

I know NVIDIA video games have some issues with Windows 10 so I wouldn't be surprised if they was the cause of the bug before.
